File: backend/app/seed.py
# Port de src/api/mock/seed.ts. Provee el "universo de lectura" (workspaces,
# datasets, tablas) para el modo data_source="seed", y los schedules iniciales
# con los que arranca el store si todavía no hay archivo de persistencia.
from .frequency import LAST_DAY, schedule_time
from .models import Dataset, LastRun, Schedule, TableInfo, Workspace

_WORKSPACES = [
    Workspace(id="ws-ventas", name="Ventas y Comercial"),
    Workspace(id="ws-finanzas", name="Finanzas"),
    Workspace(id="ws-ops", name="Operaciones"),
]

_DATASET_DEFS = [
    ("ds-ventas-retail", "Ventas Retail", "ws-ventas",
     ["Ventas", "Clientes", "Productos", "Sucursales", "Vendedores", "Promociones"]),
    ("ds-ventas-may", "Ventas Mayorista", "ws-ventas",
     ["Pedidos", "Cuentas", "Productos", "Descuentos"]),
    ("ds-pipeline", "Pipeline Comercial", "ws-ventas",
     ["Oportunidades", "Etapas", "Contactos", "Actividades", "Forecast"]),
    ("ds-pyl", "P&L Consolidado", "ws-finanzas",
     ["Asientos", "CentrosDeCosto", "Cuentas", "Periodos", "Presupuesto"]),
    ("ds-tesoreria", "Tesorería", "ws-finanzas",
     ["FlujoCaja", "Bancos", "Conciliaciones", "Pagos"]),
    ("ds-logistica", "Logística", "ws-ops",
     ["Envios", "Rutas", "Transportistas", "Entregas", "Costos", "Zonas", "Devoluciones"]),
    ("ds-inventario", "Inventario", "ws-ops",
     ["Stock", "Movimientos", "Depositos", "Lotes", "Ajustes"]),
    ("ds-rrhh", "RRHH", "ws-ops",
     ["Empleados", "Liquidaciones", "Ausencias", "Capacitaciones", "Legajos", "Evaluaciones"]),
    ("ds-calidad", "Calidad", "ws-ops",
     ["Inspecciones", "Defectos", "Auditorias", "NoConformidades"]),
]


def seed_workspaces() -> list[Workspace]:
    return [w.model_copy(deep=True) for w in _WORKSPACES]


def seed_datasets() -> list[Dataset]:
    return [
        Dataset(id=did, name=name, workspace_id=ws)
        for did, name, ws, _tables in _DATASET_DEFS
    ]


def seed_tables() -> list[TableInfo]:
    out: list[TableInfo] = []
    for did, _name, _ws, tables in _DATASET_DEFS:
        for t in tables:
            out.append(TableInfo(name=t, dataset_id=did))
    return out


def _sched(**kwargs) -> Schedule:
    # Construimos con time="" para que Pydantic convierta el dict `frequency` en su
    # modelo de unión, y recién ahí derivamos el time canónico (espejo denormalizado).
    sch = Schedule(time="", **kwargs)
    sch.time = schedule_time(sch.frequency)
    return sch


def seed_schedules() -> list[Schedule]:
    return [
        _sched(id="sch-1", dataset_id="ds-ventas-retail", workspace_id="ws-ventas",
               tables=["Ventas", "Clientes"], frequency={"kind": "daily", "time": "06:00"},
               refresh_type="full", enabled=True,
               last_run=LastRun(status="Completed", timestamp="2026-06-05T06:00:00-03:00")),
        _sched(id="sch-2", dataset_id="ds-ventas-retail", workspace_id="ws-ventas",
               tables=["Productos"], frequency={"kind": "hourly", "every_hours": 4},
               refresh_type="dataOnly", enabled=True,
               last_run=LastRun(status="InProgress", timestamp="2026-06-05T12:00:00-03:00")),
        _sched(id="sch-3", dataset_id="ds-pyl", workspace_id="ws-finanzas",
               tables=["Asientos", "CentrosDeCosto"],
               frequency={"kind": "monthly", "day_of_month": 1, "time": "01:00"},
               refresh_type="full", enabled=True,
               last_run=LastRun(status="Failed", timestamp="2026-06-01T01:00:00-03:00")),
        _sched(id="sch-4", dataset_id="ds-logistica", workspace_id="ws-ops",
               tables=["Envios"],
               frequency={"kind": "weekly", "days_of_week": [1, 3, 5], "time": "07:30"},
               refresh_type="calculate", enabled=True,
               last_run=LastRun(status="Completed", timestamp="2026-06-04T07:30:00-03:00")),
        _sched(id="sch-5", dataset_id="ds-inventario", workspace_id="ws-ops",
               tables=["Stock", "Movimientos", "Depositos"],
               frequency={"kind": "monthly", "day_of_month": LAST_DAY, "time": "23:00"},
               refresh_type="full", enabled=False,
               last_run=LastRun(status="Completed", timestamp="2026-05-31T23:00:00-03:00")),
        _sched(id="sch-6", dataset_id="ds-tesoreria", workspace_id="ws-finanzas",
               tables=["FlujoCaja"], frequency={"kind": "daily", "time": "08:00"},
               refresh_type="dataOnly", enabled=True,
               last_run=LastRun(status="InProgress", timestamp="2026-06-05T08:00:00-03:00")),
    ]

File: src/api/mock/seed.ts
// Datos inventados iniciales: 3 workspaces, 2-4 modelos por workspace,
// 4-8 tablas por modelo, y algunos schedules preexistentes con distintos
// estados de último run (Completado / Falló / En curso).
import type { Schedule, TableInfo } from '../types'
import { LAST_DAY, scheduleTime } from '../../domain/frequency'
import type { MockDb } from './store'

/** Completa Schedule.time a partir de la frecuencia (espejo denormalizado). */
function sched(s: Omit<Schedule, 'time'>): Schedule {
  return { ...s, time: scheduleTime(s.frequency) }
}

export function createSeedDb(): MockDb {
  const workspaces = [
    { id: 'ws-ventas', name: 'Ventas y Comercial' },
    { id: 'ws-finanzas', name: 'Finanzas' },
    { id: 'ws-ops', name: 'Operaciones' },
  ]

  const datasetDefs: Array<{
    id: string
    name: string
    workspaceId: string
    tables: string[]
  }> = [
    { id: 'ds-ventas-retail', name: 'Ventas Retail', workspaceId: 'ws-ventas', tables: ['Ventas', 'Clientes', 'Productos', 'Sucursales', 'Vendedores', 'Promociones'] },
    { id: 'ds-ventas-may', name: 'Ventas Mayorista', workspaceId: 'ws-ventas', tables: ['Pedidos', 'Cuentas', 'Productos', 'Descuentos'] },
    { id: 'ds-pipeline', name: 'Pipeline Comercial', workspaceId: 'ws-ventas', tables: ['Oportunidades', 'Etapas', 'Contactos', 'Actividades', 'Forecast'] },
    { id: 'ds-pyl', name: 'P&L Consolidado', workspaceId: 'ws-finanzas', tables: ['Asientos', 'CentrosDeCosto', 'Cuentas', 'Periodos', 'Presupuesto'] },
    { id: 'ds-tesoreria', name: 'Tesorería', workspaceId: 'ws-finanzas', tables: ['FlujoCaja', 'Bancos', 'Conciliaciones', 'Pagos'] },
    { id: 'ds-logistica', name: 'Logística', workspaceId: 'ws-ops', tables: ['Envios', 'Rutas', 'Transportistas', 'Entregas', 'Costos', 'Zonas', 'Devoluciones'] },
    { id: 'ds-inventario', name: 'Inventario', workspaceId: 'ws-ops', tables: ['Stock', 'Movimientos', 'Depositos', 'Lotes', 'Ajustes'] },
    { id: 'ds-rrhh', name: 'RRHH', workspaceId: 'ws-ops', tables: ['Empleados', 'Liquidaciones', 'Ausencias', 'Capacitaciones', 'Legajos', 'Evaluaciones'] },
    { id: 'ds-calidad', name: 'Calidad', workspaceId: 'ws-ops', tables: ['Inspecciones', 'Defectos', 'Auditorias', 'NoConformidades'] },
  ]

  const datasets = datasetDefs.map(({ id, name, workspaceId }) => ({ id, name, workspaceId }))
  const tables: TableInfo[] = datasetDefs.flatMap((d) =>
    d.tables.map((name) => ({ name, datasetId: d.id })),
  )

  const schedules: Schedule[] = [
    sched({ id: 'sch-1', datasetId: 'ds-ventas-retail', workspaceId: 'ws-ventas', tables: ['Ventas', 'Clientes'], frequency: { kind: 'daily', time: '06:00' }, refreshType: 'full', enabled: true, lastRun: { status: 'Completed', timestamp: '2026-06-05T06:00:00-03:00' } }),
    sched({ id: 'sch-2', datasetId: 'ds-ventas-retail', workspaceId: 'ws-ventas', tables: ['Productos'], frequency: { kind: 'hourly', everyHours: 4 }, refreshType: 'dataOnly', enabled: true, lastRun: { status: 'InProgress', timestamp: '2026-06-05T12:00:00-03:00' } }),
    sched({ id: 'sch-3', datasetId: 'ds-pyl', workspaceId: 'ws-finanzas', tables: ['Asientos', 'CentrosDeCosto'], frequency: { kind: 'monthly', dayOfMonth: 1, time: '01:00' }, refreshType: 'full', enabled: true, lastRun: { status: 'Failed', timestamp: '2026-06-01T01:00:00-03:00' } }),
    sched({ id: 'sch-4', datasetId: 'ds-logistica', workspaceId: 'ws-ops', tables: ['Envios'], frequency: { kind: 'weekly', daysOfWeek: [3], time: '07:30' }, refreshType: 'calculate', enabled: true, lastRun: { status: 'Completed', timestamp: '2026-06-04T07:30:00-03:00' } }),
    sched({ id: 'sch-5', datasetId: 'ds-inventario', workspaceId: 'ws-ops', tables: ['Stock', 'Movimientos', 'Depositos'], frequency: { kind: 'monthly', dayOfMonth: LAST_DAY, time: '23:00' }, refreshType: 'full', enabled: false, lastRun: { status: 'Completed', timestamp: '2026-05-31T23:00:00-03:00' } }),
    sched({ id: 'sch-6', datasetId: 'ds-tesoreria', workspaceId: 'ws-finanzas', tables: ['FlujoCaja'], frequency: { kind: 'daily', time: '08:00', daysOfWeek: [1, 2, 3, 4, 5] }, refreshType: 'dataOnly', enabled: true, lastRun: { status: 'InProgress', timestamp: '2026-06-05T08:00:00-03:00' } }),
  ]

  // Sincronizar scheduleId en las tablas referenciadas por cada schedule.
  for (const sch of schedules) {
    for (const name of sch.tables) {
      const t = tables.find((tb) => tb.datasetId === sch.datasetId && tb.name === name)
      if (t) t.scheduleId = sch.id
    }
  }

  return { workspaces, datasets, tables, schedules }
}
